src/FakeItEasy/Core/FakeManager.AutoFakePropertyRule.cs

```diff
@@ -29,7 +29,12 @@ public override void Apply(IInterceptedFakeObjectCall fakeObjectCall)
                                       CalledNumberOfTimes = 1
                                   };
 
-                Fake.GetFakeManager(fakeObjectCall.FakedObject).allUserRules.AddFirst(newRule);
+                var userRules = Fake.GetFakeManager(fakeObjectCall.FakedObject).allUserRules;
+                lock (userRules)
+                {
+                    userRules.AddFirst(newRule);
+                }
+

```

Comment:
This makes a lot of sense. I'm happy to do so, but the PropertySetterRule [case](https://github.com/FakeItEasy/FakeItEasy/pull/1571/files#diff-83eba6f3619b32e112e4ed50efd6be87R24) is quite a bit more complicated. Would you like a `AddRuleLastIfThereIsNotAlreadyAMatchingSetterRule`?

Reply:
> I'm happy to do so, but the PropertySetterRule [case](https://github.com/FakeItEasy/FakeItEasy/pull/1571/files#diff-83eba6f3619b32e112e4ed50efd6be87R24) is quite a bit more complicated. Would you like a `AddRuleLastIfThereIsNotAlreadyAMatchingSetterRule`?

No, that would probably be overly specific... Maybe just leave it as is, then.

Reply:
I thought about this last night. An alternative would be a method that takes an Action that would mutate the list. The action would be called once we lock access to the list. This could be used by both `PropertySetterRule` and `AutoFakePropertyRule`.

Reply:
In theory, this method could be used throughout `FakeManager`. That seems a little too abstract, I thought, but if you like the approach either for the nested classes or throughout, I can implement.

Reply:
I like the idea!

Reply:
Before I start typing, like it for the nested classes only, or throughout? I'd hate to disappoint you.

Reply:
Pushing a change that just affects the nested classes for now, as it's easier than undoing. And I slightly prefer this version anyhow.

Reply:
For consistency, maybe it would be better to do it everywhere, but I don't mind either way.

Reply:
I just gave it a go, and I didn't really enjoy the result. It looks funny to use a `MutateUserRules` to just fetch information (like for the `Rules` property) as a side effect. Alternatively, we could have a method that takes a `Func` for those operations, or exempt them, but that just seems to add more complexity. I think "lock if you're a `FakeManager` member and use the mutating function otherwise" is simple enough to work and doesn't make the rest o the code too gross. Since you don't mind either way, let's just leave it. I did notice that I could've made the method private (or could leave it `internal` but I have to move it to appease StyleCop),  so I'll push the fixup for your enjoyment.

Reply:
Fine by me 👍